### DISCUSSION TITLE
Proposal: implicitly typed array creation expression with size

### DISCUSSION THREAD

**Discussion by Tan90909090** [dotnet/csharplang#683] [General]
## Background
We have several ways to create array.

Firstly, we can specify array size:
```cs
var withSize = new int[3] { 1, 2, 3 };
```
Compiler detects a mismatch between specified size and array-initializer's length and reports an error:
```cs
// compile error
var mismatch = new int[1] { 1, 2, 3 };
```

Secondly, we can omit element type. Compiler infers element type:
```cs
var inferred = new[] { 1, 2, 3 };
```
## Problem
Currently, we can not combine both features:
```cs
// syntax error
var ng = new[3] { 1, 2, 3 };
```
## Proposal
Let's allow implicitly typed array creation expression with size.
Syntax may change like this:
```diff
array_creation_expression
    : 'new' non_array_type '[' expression_list ']' rank_specifier* array_initializer?
    | 'new' array_type array_initializer
    | 'new' rank_specifier array_initializer
+   | 'new' '[' expression_list ']' array_initializer
    ;
```

**Comment by jnm2** [comment]
> #### Problem
> 
> Currently, we can not combine both features

Is this a problem? I think this is a feature, not a bug.

**Comment by yaakov-h** [comment]
> Let's allow implicitly typed array creation expression with size.

Why?

**Comment by brandon942** [comment]
Why not javascript like simplicity with type and size inference?  

    var a = [1d, 2, 3, 4, 5, 6]; // let first element determine type
